File: src/main/java/com/tkr/desk/HaircutCalculator.java
package com.tkr.desk;

import com.tkr.types.WireTypes.Status;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Tiered collateral haircut grid by asset class, rating, and concentration.
 * Ported from legacy-cpp haircut_calculator.cc.
 */
public final class HaircutCalculator {

    public enum AssetClass {
        EQUITY, CORPORATE_BOND, GOVERNMENT_BOND, MUNICIPAL_BOND, PREFERRED, CONVERTIBLE, OTHER
    }

    public enum CreditRating {
        AAA, AA, A, BBB, BB, B, NOT_RATED
    }

    public enum CollateralTier { TIER1, TIER2, TIER3, ILLIQUID }

    public static final class HaircutGridCell {
        public AssetClass assetClass;
        public CreditRating rating;
        public int haircutBp;
        public int concentrationCapBp = 1000;
    }

    public static final class HaircutInput {
        public int symbolId;
        public AssetClass assetClass = AssetClass.EQUITY;
        public CreditRating rating = CreditRating.A;
        public long marketValueCents;
        public int portfolioWeightBp;
    }

    public static final class HaircutResult {
        public int symbolId;
        public long grossValueCents;
        public int effectiveHaircutBp;
        public long haircutAmountCents;
        public long collateralValueCents;
    }

    public static final class HaircutCalculatorConfig {
        public int deskId;
        public boolean applyConcentrationPenalty;

        public HaircutCalculatorConfig(int deskId, boolean applyConcentrationPenalty) {
            this.deskId = deskId;
            this.applyConcentrationPenalty = applyConcentrationPenalty;
        }
    }

    public static final class HaircutCalculatorSummary {
        public Status status = Status.OK;
        public List<HaircutResult> results = new ArrayList<>();
        public long totalGrossCents;
        public long totalHaircutCents;
        public long totalCollateralCents;
    }

    public static final class HaircutTierRow {
        public long notionalThresholdCents;
        public int haircutBp;
    }

    private static final class ValueTier {
        long valueFloorCents;
        long valueCeilingCents;
        int additionalHaircutBp;
    }

    private static final ValueTier[] EQUITY_TIERS = {
        tier(0, 10_000_000L, 0),
        tier(10_000_000L, 50_000_000L, 200),
        tier(50_000_000L, 200_000_000L, 500),
        tier(200_000_000L, Long.MAX_VALUE, 1000)
    };

    private final HaircutCalculatorConfig config;
    private final Map<Long, HaircutGridCell> grid = new HashMap<>();
    private final CollateralTier[] symbolTier = new CollateralTier[100000];
    private final HaircutTierRow[][] tierGrid = new HaircutTierRow[4][];

    public HaircutCalculator() {
        this(new HaircutCalculatorConfig(0, true));
    }

    public HaircutCalculator(HaircutCalculatorConfig config) {
        this.config = config != null ? config : new HaircutCalculatorConfig(0, true);
        loadDefaultGrid();
        tierGrid[0] = new HaircutTierRow[] { row(0, 0), row(1_000_000_00L, 200), row(10_000_000_00L, 500) };
        tierGrid[1] = new HaircutTierRow[] { row(0, 500), row(1_000_000_00L, 1000), row(10_000_000_00L, 1500) };
        tierGrid[2] = new HaircutTierRow[] { row(0, 1500), row(500_000_00L, 2000), row(5_000_000_00L, 2500) };
        tierGrid[3] = new HaircutTierRow[] { row(0, 3000), row(100_000_00L, 4000), row(1_000_000_00L, 5000) };
        symbolTier[42] = CollateralTier.TIER1;
    }

    private static ValueTier tier(long floor, long ceiling, int bp) {
        ValueTier t = new ValueTier();
        t.valueFloorCents = floor;
        t.valueCeilingCents = ceiling;
        t.additionalHaircutBp = bp;
        return t;
    }

    private static HaircutTierRow row(long threshold, int bp) {
        HaircutTierRow r = new HaircutTierRow();
        r.notionalThresholdCents = threshold;
        r.haircutBp = bp;
        return r;
    }

    private static long gridKey(AssetClass assetClass, CreditRating rating) {
        return ((long) assetClass.ordinal() << 8) | rating.ordinal();
    }

    public void setGridCell(HaircutGridCell cell) {
        grid.put(gridKey(cell.assetClass, cell.rating), cell);
    }

    public static int lookupBaseHaircutBp(AssetClass assetClass, CreditRating rating) {
        return switch (assetClass) {
            case EQUITY -> 1500;
            case CORPORATE_BOND -> switch (rating) {
                case AAA -> 200;
                case AA -> 400;
                case A -> 600;
                case BBB -> 1000;
                case BB -> 1500;
                case B -> 2000;
                default -> 2500;
            };
            case GOVERNMENT_BOND -> switch (rating) {
                case AAA -> 0;
                case AA -> 50;
                default -> 100;
            };
            case MUNICIPAL_BOND -> 500;
            case PREFERRED -> 1200;
            case CONVERTIBLE -> 1800;
            default -> 2500;
        };
    }

    public static int applyConcentrationPenalty(int baseHaircutBp, int weightBp) {
        if (weightBp <= 1000) return baseHaircutBp;
        int excessBp = weightBp - 1000;
        return baseHaircutBp + (excessBp * 50) / 100;
    }

    private static int lookupTierPenalty(long marketValueCents) {
        for (ValueTier tier : EQUITY_TIERS) {
            if (marketValueCents >= tier.valueFloorCents && marketValueCents < tier.valueCeilingCents) {
                return tier.additionalHaircutBp;
            }
        }
        return 0;
    }

    public void loadDefaultGrid() {
        grid.clear();
        for (AssetClass ac : AssetClass.values()) {
            for (CreditRating rt : CreditRating.values()) {
                HaircutGridCell cell = new HaircutGridCell();
                cell.assetClass = ac;
                cell.rating = rt;
                cell.haircutBp = lookupBaseHaircutBp(ac, rt);
                setGridCell(cell);
            }
        }
    }

    public int lookupGridHaircut(AssetClass assetClass, CreditRating rating) {
        HaircutGridCell cell = grid.get(gridKey(assetClass, rating));
        return cell != null ? cell.haircutBp : lookupBaseHaircutBp(assetClass, rating);
    }

    public HaircutResult computeSingle(HaircutInput input) {
        HaircutResult result = new HaircutResult();
        result.symbolId = input.symbolId;
        result.grossValueCents = input.marketValueCents;
        int haircutBp = lookupGridHaircut(input.assetClass, input.rating);
        haircutBp += lookupTierPenalty(input.marketValueCents);
        if (config.applyConcentrationPenalty) {
            HaircutGridCell cell = grid.get(gridKey(input.assetClass, input.rating));
            if (cell != null && input.portfolioWeightBp > cell.concentrationCapBp) {
                haircutBp = applyConcentrationPenalty(haircutBp, input.portfolioWeightBp);
            }
        }
        if (haircutBp > 10000) haircutBp = 10000;
        result.effectiveHaircutBp = haircutBp;
        result.haircutAmountCents = input.marketValueCents * haircutBp / 10000L;
        result.collateralValueCents = input.marketValueCents - result.haircutAmountCents;
        return result;
    }

    public HaircutCalculatorSummary compute(List<HaircutInput> inputs) {
        HaircutCalculatorSummary summary = new HaircutCalculatorSummary();
        summary.status = Status.OK;
        if (inputs == null || inputs.isEmpty()) return summary;
        long totalGross = 0;
        for (HaircutInput input : inputs) totalGross += input.marketValueCents;
        for (HaircutInput input : inputs) {
            HaircutInput adjusted = input;
            if (adjusted.portfolioWeightBp == 0 && totalGross > 0) {
                adjusted.portfolioWeightBp = (int) (input.marketValueCents * 10000L / totalGross);
            }
            HaircutResult result = computeSingle(adjusted);
            summary.results.add(result);
            summary.totalGrossCents += result.grossValueCents;
            summary.totalHaircutCents += result.haircutAmountCents;
            summary.totalCollateralCents += result.collateralValueCents;
        }
        summary.results.sort(Comparator.comparingInt((HaircutResult r) -> r.effectiveHaircutBp).reversed());
        return summary;
    }

    public void assignTier(int symbolId, CollateralTier tier) {
        if (symbolId >= 0 && symbolId < symbolTier.length) symbolTier[symbolId] = tier;
    }

    public int lookupHaircutBp(int symbolId, long notionalCents) {
        CollateralTier tier = symbolId >= 0 && symbolId < symbolTier.length && symbolTier[symbolId] != null
                ? symbolTier[symbolId] : CollateralTier.TIER2;
        HaircutTierRow[] rows = tierGrid[tier.ordinal()];
        int bp = rows[0].haircutBp;
        for (HaircutTierRow row : rows) {
            if (notionalCents >= row.notionalThresholdCents) bp = row.haircutBp;
        }
        return bp;
    }

    public long computeHaircutCents(int symbolId, long notionalCents) {
        if (notionalCents < 0) return -1;
        return notionalCents * lookupHaircutBp(symbolId, notionalCents) / 10000L;
    }

    public long collateralValueCents(int symbolId, long notionalCents) {
        return notionalCents - computeHaircutCents(symbolId, notionalCents);
    }

    public int[] buildHaircutCurve(int symbolId, long[] notionals) {
        int[] out = new int[notionals.length];
        for (int i = 0; i < notionals.length; i++) out[i] = lookupHaircutBp(symbolId, notionals[i]);
        return out;
    }

    public HaircutCalculatorSummary computeFromBatch(com.tkr.types.WireTypes.BatchWireFrame frame) {
        List<HaircutInput> inputs = new ArrayList<>();
        long totalGross = 0;
        for (com.tkr.types.WireTypes.WireBatchRecord rec : frame.records) {
            totalGross += (long) rec.qtyMilli * rec.priceTick / 1000L;
        }
        for (com.tkr.types.WireTypes.WireBatchRecord rec : frame.records) {
            HaircutInput input = new HaircutInput();
            input.symbolId = rec.symbolId;
            input.marketValueCents = (long) rec.qtyMilli * rec.priceTick / 1000L;
            if (totalGross > 0) input.portfolioWeightBp = (int) (input.marketValueCents * 10000L / totalGross);
            inputs.add(input);
        }
        return compute(inputs);
    }
}

File: legacy-cpp/src/desk/haircut_calculator.cc
#include "tkr/desk/haircut_calculator.h"

#include <algorithm>
#include <climits>
#include <unordered_map>

namespace tkr {
namespace desk {
namespace {

struct HaircutTier {
  std::int64_t value_floor_cents;
  std::int64_t value_ceiling_cents;
  std::uint32_t additional_haircut_bp;
};

const HaircutTier kEquityTiers[] = {
    {0, 10000000, 0},
    {10000000, 50000000, 200},
    {50000000, 200000000, 500},
    {200000000, INT64_MAX, 1000},
};

std::uint32_t LookupTierPenalty(std::int64_t market_value_cents) {
  for (const HaircutTier& tier : kEquityTiers) {
    if (market_value_cents >= tier.value_floor_cents &&
        market_value_cents < tier.value_ceiling_cents) {
      return tier.additional_haircut_bp;
    }
  }
  return 0;
}

}  // namespace

HaircutCalculator::HaircutCalculator(HaircutCalculatorConfig config)
    : config_(config) {
  LoadDefaultGrid();
}

std::uint64_t HaircutCalculator::GridKey(AssetClass asset_class,
                                         CreditRating rating) const {
  return (static_cast<std::uint64_t>(asset_class) << 8) |
         static_cast<std::uint64_t>(rating);
}

void HaircutCalculator::SetGridCell(const HaircutGridCell& cell) {
  grid_[GridKey(cell.asset_class, cell.rating)] = cell;
}

std::uint32_t HaircutCalculator::LookupBaseHaircutBp(AssetClass asset_class,
                                                     CreditRating rating) {
  switch (asset_class) {
    case AssetClass::kEquity:
      return 1500;
    case AssetClass::kCorporateBond:
      switch (rating) {
        case CreditRating::kAaa:
          return 200;
        case CreditRating::kAa:
          return 400;
        case CreditRating::kA:
          return 600;
        case CreditRating::kBbb:
          return 1000;
        case CreditRating::kBb:
          return 1500;
        case CreditRating::kB:
          return 2000;
        default:
          return 2500;
      }
    case AssetClass::kGovernmentBond:
      switch (rating) {
        case CreditRating::kAaa:
          return 0;
        case CreditRating::kAa:
          return 50;
        default:
          return 100;
      }
    case AssetClass::kMunicipalBond:
      return 500;
    case AssetClass::kPreferred:
      return 1200;
    case AssetClass::kConvertible:
      return 1800;
    default:
      return 2500;
  }
}

std::uint32_t HaircutCalculator::ApplyConcentrationPenalty(
    std::uint32_t base_haircut_bp, std::uint32_t weight_bp) {
  if (weight_bp <= 1000) {
    return base_haircut_bp;
  }
  const std::uint32_t excess_bp = weight_bp - 1000;
  const std::uint32_t penalty = (excess_bp * 50) / 100;
  return base_haircut_bp + penalty;
}

void HaircutCalculator::LoadDefaultGrid() {
  for (std::uint8_t ac = 0; ac <= static_cast<std::uint8_t>(AssetClass::kConvertible);
       ++ac) {
    for (std::uint8_t rt = 0; rt <= static_cast<std::uint8_t>(CreditRating::kNotRated);
         ++rt) {
      HaircutGridCell cell{};
      cell.asset_class = static_cast<AssetClass>(ac);
      cell.rating = static_cast<CreditRating>(rt);
      cell.haircut_bp = LookupBaseHaircutBp(cell.asset_class, cell.rating);
      cell.concentration_cap_bp = 1000;
      SetGridCell(cell);
    }
  }
}

std::uint32_t HaircutCalculator::LookupGridHaircut(
    AssetClass asset_class, CreditRating rating) const {
  const auto it = grid_.find(GridKey(asset_class, rating));
  if (it == grid_.end()) {
    return LookupBaseHaircutBp(asset_class, rating);
  }
  return it->second.haircut_bp;
}

HaircutResult HaircutCalculator::ComputeSingle(const HaircutInput& input) const {
  HaircutResult result{};
  result.symbol_id = input.symbol_id;
  result.gross_value_cents = input.market_value_cents;

  std::uint32_t haircut_bp = LookupGridHaircut(input.asset_class, input.rating);

  const std::uint32_t tier_penalty =
      LookupTierPenalty(input.market_value_cents);
  haircut_bp += tier_penalty;

  if (config_.apply_concentration_penalty) {
    const auto it = grid_.find(GridKey(input.asset_class, input.rating));
    if (it != grid_.end() &&
        input.portfolio_weight_bp > it->second.concentration_cap_bp) {
      haircut_bp = ApplyConcentrationPenalty(haircut_bp, input.portfolio_weight_bp);
    }
  }

  if (haircut_bp > 10000) {
    haircut_bp = 10000;
  }

  result.effective_haircut_bp = haircut_bp;
  result.haircut_amount_cents =
      (input.market_value_cents * static_cast<std::int64_t>(haircut_bp)) / 10000;
  result.collateral_value_cents =
      input.market_value_cents - result.haircut_amount_cents;

  return result;
}

HaircutCalculatorSummary HaircutCalculator::Compute(
    const std::vector<HaircutInput>& inputs) {
  HaircutCalculatorSummary summary{};
  summary.status = Status::kOk;

  if (inputs.empty()) {
    return summary;
  }

  summary.results.reserve(inputs.size());

  std::int64_t total_gross = 0;
  for (const HaircutInput& input : inputs) {
    total_gross += input.market_value_cents;
  }

  for (const HaircutInput& input : inputs) {
    HaircutInput adjusted = input;
    if (adjusted.portfolio_weight_bp == 0 && total_gross > 0) {
      adjusted.portfolio_weight_bp = static_cast<std::uint32_t>(
          (input.market_value_cents * 10000) / total_gross);
    }

    HaircutResult result = ComputeSingle(adjusted);
    summary.results.push_back(result);
    summary.total_gross_cents += result.gross_value_cents;
    summary.total_haircut_cents += result.haircut_amount_cents;
    summary.total_collateral_cents += result.collateral_value_cents;
  }

  std::sort(summary.results.begin(), summary.results.end(),
            [](const HaircutResult& a, const HaircutResult& b) {
              return a.effective_haircut_bp > b.effective_haircut_bp;
            });

  return summary;
}

}  // namespace desk
}  // namespace tkr
